grouper/models/public_key.py

```diff
@@ -2,6 +2,63 @@
 from sqlalchemy import Column, Integer, String, Text, ForeignKey, DateTime
 from sqlalchemy.orm import relationship
 from grouper.models.base.model_base import Model
+from grouper.models.public_key_tag_map import PublicKeyTagMap
+
+
+def _perm_list_to_dict(perms):
+    # type: List[Permission] -> Dict[str, Dict[str, Permission]]
+    """Converts a list of Permission objects into a dictionary indexed by the permission names.
+    That dictionary in turn holds another dictionary which is indexed by permission argument, and
+    stores the Permission object
+
+    Args:
+        perms: a list containing Permission objects
+
+    Returns:
+        a dictionary with the permission names as keys, and has as values another dictionary
+        with permission arguments as keys and Permission objects as values
+    """
+    ret = dict()
+    for perm in perms:
+        if perm.name not in ret:
+            ret[perm.name] = dict()
+        ret[perm.name][perm.argument] = perm
+    return ret
+
+
+def _permission_intersection(perms_a, perms_b):
+    # type: List[Permission], List[Permission] -> Set[Permission]
+    """Returns the intersection of the two Permission lists, taking into account the special
+    handling of argument wildcards
+
+    Args:
+        perms_a: the first list of permissions
+        perms_b: the second list of permissions
+
+    Returns:
+        a set of all permissions that both perms_a and perms_b grant access to
+    """
+    pdict_b = _perm_list_to_dict(perms_b)
+    ret = set()
+    for perm in perms_a:
+        if perm.name not in pdict_b:
+            continue
+        if perm.argument in pdict_b[perm.name]:
+            ret.add(perm)
+            continue
+        # Argument wildcard
+        if "*" in pdict_b[perm.name]:
+            ret.add(perm)
+            continue
+        # According to Group.has_permission, None as an argument counts as a wildcard too
+        if None in pdict_b[perm.name]:
+            ret.add(perm)
+            continue
+        # If this permission is a wildcard, we add all permissions with the same name from
+        # the other set
+        if perm.argument == "*" or perm.argument is None:
+            ret |= {p for p in pdict_b[perm.name].values()}
+    return ret
```

Comment:
I think I get why you put these guys here (they're only used within this module) but they feel like they belong with the rest of permission code since it's whole-y to do with permissions. Thoughts on putting this in `grouper.permissions` or some such?